File: go/config/container/container_test.go
package container

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	config "github.com/SmooAI/config/go/config"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// stubClient is an in-memory ConfigClient for the injection-seam tests. It
// records the env-tier seed and serves values without a network round-trip.
type stubClient struct {
	mu          sync.Mutex
	values      map[string]any
	cache       map[string]any
	getValueErr error // when set, GetValue returns this
	allValueErr error // when set, GetAllValues returns this
	getValueN   int32
}

func newStubClient(values map[string]any) *stubClient {
	return &stubClient{values: values, cache: map[string]any{}}
}

func (s *stubClient) GetAllValues(environment string) (map[string]any, error) {
	if s.allValueErr != nil {
		return nil, s.allValueErr
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	for k, v := range s.values {
		s.cache[k] = v
	}
	out := map[string]any{}
	for k, v := range s.values {
		out[k] = v
	}
	return out, nil
}

func (s *stubClient) GetValue(key, environment string) (any, error) {
	atomic.AddInt32(&s.getValueN, 1)
	if s.getValueErr != nil {
		return nil, s.getValueErr
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	v, ok := s.values[key]
	if !ok {
		return nil, nil
	}
	s.cache[key] = v
	return v, nil
}

func (s *stubClient) GetCachedValue(key, environment string) (any, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	v, ok := s.cache[key]
	return v, ok
}

func (s *stubClient) SeedCache(key string, value any, environment string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.cache[key] = value
}

func testSchema(t *testing.T) *config.ConfigDefinition {
	t.Helper()
	return config.DefineConfig(
		map[string]any{"type": "object", "properties": map[string]any{"apiUrl": map[string]any{"type": "string"}}},
		map[string]any{"type": "object", "properties": map[string]any{"stripeApiKey": map[string]any{"type": "string"}}},
		map[string]any{"type": "object", "properties": map[string]any{"betaEnabled": map[string]any{"type": "boolean"}}},
	)
}

// --- §7.2: bootstrap-missing-env errors + lists missing -----------------------

func TestInitContainerConfig_MissingEnv_ErrorsAndListsMissing(t *testing.T) {
	_, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema:      testSchema(t),
		EnvOverride: map[string]string{}, // nothing set
	})
	require.Error(t, err)

	var be *ConfigBootstrapError
	require.True(t, errors.As(err, &be), "expected *ConfigBootstrapError, got %T", err)
	assert.ElementsMatch(t, []string{
		"SMOOAI_CONFIG_API_URL",
		"SMOOAI_CONFIG_CLIENT_ID",
		"SMOOAI_CONFIG_CLIENT_SECRET",
		"SMOOAI_CONFIG_ORG_ID",
		"SMOOAI_CONFIG_ENV",
	}, be.Missing)
	assert.Contains(t, be.Error(), "SMOOAI_CONFIG_CLIENT_ID")
}

func TestInitContainerConfig_BlankEnvCountsAsMissing(t *testing.T) {
	_, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema: testSchema(t),
		EnvOverride: map[string]string{
			"SMOOAI_CONFIG_API_URL":       "  ", // blank
			"SMOOAI_CONFIG_CLIENT_ID":     "id",
			"SMOOAI_CONFIG_CLIENT_SECRET": "sk",
			"SMOOAI_CONFIG_ORG_ID":        "org",
			"SMOOAI_CONFIG_ENV":           "production",
		},
	})
	var be *ConfigBootstrapError
	require.ErrorAs(t, err, &be)
	assert.Equal(t, []string{"SMOOAI_CONFIG_API_URL"}, be.Missing)
}

func TestInitContainerConfig_MissingSchema(t *testing.T) {
	_, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{})
	var be *ConfigBootstrapError
	require.ErrorAs(t, err, &be)
	assert.Equal(t, []string{"Schema"}, be.Missing)
}

func TestInitContainerConfig_InjectedClient_OnlyEnvRequired(t *testing.T) {
	// With a ConfigClient injected, only SMOOAI_CONFIG_ENV is required.
	stub := newStubClient(map[string]any{"stripeApiKey": "sk_live_x"})
	h, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema:       testSchema(t),
		ConfigClient: stub,
		EnvOverride:  map[string]string{"SMOOAI_CONFIG_ENV": "production"},
	})
	require.NoError(t, err)
	require.NotNil(t, h)
	assert.Equal(t, "production", h.environment)
}

func TestInitContainerConfig_InjectedClient_MissingEnvStillErrors(t *testing.T) {
	stub := newStubClient(map[string]any{})
	_, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema:       testSchema(t),
		ConfigClient: stub,
		EnvOverride:  map[string]string{},
	})
	var be *ConfigBootstrapError
	require.ErrorAs(t, err, &be)
	assert.Equal(t, []string{"SMOOAI_CONFIG_ENV"}, be.Missing)
}

// --- §7.2: happy path fetch + cache ------------------------------------------

func newHandle(t *testing.T, values map[string]any, opts ...func(*InitContainerConfigOptions)) (*ContainerConfigHandle, *stubClient) {
	t.Helper()
	stub := newStubClient(values)
	o := InitContainerConfigOptions{
		Schema:       testSchema(t),
		ConfigClient: stub,
		EnvOverride:  map[string]string{"SMOOAI_CONFIG_ENV": "production"},
	}
	for _, f := range opts {
		f(&o)
	}
	h, err := InitContainerConfig(context.Background(), o)
	require.NoError(t, err)
	return h, stub
}

func TestGet_HappyPath(t *testing.T) {
	h, _ := newHandle(t, map[string]any{"stripeApiKey": "sk_live_x", "apiUrl": "https://api.smoo.ai"})

	v, ok, err := h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	assert.True(t, ok)
	assert.Equal(t, "sk_live_x", v)

	pv, ok, err := h.PublicConfig.Get("apiUrl")
	require.NoError(t, err)
	assert.True(t, ok)
	assert.Equal(t, "https://api.smoo.ai", pv)
}

func TestGet_UsesCacheAfterInitialFetch(t *testing.T) {
	h, stub := newHandle(t, map[string]any{"stripeApiKey": "sk_live_x"})
	// MustGet reads cache only — the initial GetAllValues seeded it, so no
	// GetValue network call should be needed.
	v, ok := h.SecretConfig.MustGet("stripeApiKey")
	assert.True(t, ok)
	assert.Equal(t, "sk_live_x", v)
	assert.Equal(t, int32(0), atomic.LoadInt32(&stub.getValueN), "MustGet must not call GetValue")
}

// --- §7.2: required-key-unresolved errors (not absent) -----------------------

func TestGet_RequiredKeyUnresolved_Errors(t *testing.T) {
	h, _ := newHandle(t, map[string]any{}) // server has no value for the key

	_, ok, err := h.SecretConfig.Get("stripeApiKey")
	assert.False(t, ok)
	require.Error(t, err)
	var ue *ConfigKeyUnresolvedError
	require.ErrorAs(t, err, &ue)
	assert.Equal(t, "stripeApiKey", ue.Key)
	assert.Equal(t, "production", ue.Env)
	assert.Equal(t, []Tier{TierEnv, TierHTTP}, ue.TriedTiers)
}

func TestMustGet_RequiredKeyUnresolved_Panics(t *testing.T) {
	h, _ := newHandle(t, map[string]any{})
	defer func() {
		r := recover()
		require.NotNil(t, r, "MustGet should panic on unresolved required key")
		ue, ok := r.(*ConfigKeyUnresolvedError)
		require.True(t, ok, "panic value should be *ConfigKeyUnresolvedError, got %T", r)
		assert.Equal(t, "stripeApiKey", ue.Key)
	}()
	h.SecretConfig.MustGet("stripeApiKey")
}

// --- §7.2: optional-key-absent returns absent (ok=false, no error) -----------

func TestGet_OptionalKeyAbsent_NoError(t *testing.T) {
	h, _ := newHandle(t, map[string]any{}, func(o *InitContainerConfigOptions) {
		o.OptionalKeys = []string{"stripeApiKey"}
	})
	v, ok, err := h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	assert.False(t, ok)
	assert.Nil(t, v)
}

func TestMustGet_OptionalKeyAbsent_NoPanic(t *testing.T) {
	h, _ := newHandle(t, map[string]any{}, func(o *InitContainerConfigOptions) {
		o.OptionalKeys = []string{"stripeApiKey"}
	})
	v, ok := h.SecretConfig.MustGet("stripeApiKey")
	assert.False(t, ok)
	assert.Nil(t, v)
}

func TestGet_EmptyKey_Errors(t *testing.T) {
	h, _ := newHandle(t, map[string]any{})
	_, _, err := h.SecretConfig.Get("")
	require.Error(t, err)
	assert.Contains(t, err.Error(), "empty key")
}

// --- env-tier override wins over HTTP ----------------------------------------

func TestGet_EnvOverrideWinsOverHTTP(t *testing.T) {
	h, _ := newHandle(t, map[string]any{"stripeApiKey": "from_server"}, func(o *InitContainerConfigOptions) {
		o.EnvOverride = map[string]string{
			"SMOOAI_CONFIG_ENV": "production",
			"STRIPE_API_KEY":    "from_env",
		}
	})
	v, ok, err := h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	assert.True(t, ok)
	assert.Equal(t, "from_env", v)
}

// --- §5: background-refresh failure serves last-good cached value ------------

func TestGet_RefreshFailure_ServesLastGood(t *testing.T) {
	h, stub := newHandle(t, map[string]any{"stripeApiKey": "cached_value"})
	// Prime the cache with a successful read.
	v, ok, err := h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	require.True(t, ok)
	require.Equal(t, "cached_value", v)

	// Now force GetValue to fail; the cached value should still be served.
	stub.getValueErr = errors.New("network down")
	v, ok, err = h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	assert.True(t, ok)
	assert.Equal(t, "cached_value", v)
}

// --- §7.2: 401 → refresh → retry (exercises the real ConfigClient) -----------

func Test401_Refresh_Retry(t *testing.T) {
	var tokenMints int32
	var valuesAttempts int32

	auth := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt32(&tokenMints, 1)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"access_token":"jwt-token","expires_in":3600}`))
	}))
	defer auth.Close()

	api := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		n := atomic.AddInt32(&valuesAttempts, 1)
		if n == 1 {
			// First values fetch (initial getAllValues) succeeds.
			w.Header().Set("Content-Type", "application/json")
			_, _ = w.Write([]byte(`{"values":{"stripeApiKey":"sk_live_x"}}`))
			return
		}
		// On the GetValue read, return 401 the first time, then 200.
		if n == 2 {
			w.WriteHeader(http.StatusUnauthorized)
			_, _ = w.Write([]byte(`{"error":"expired"}`))
			return
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"value":"sk_live_refreshed"}`))
	}))
	defer api.Close()

	h, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema:       testSchema(t),
		APIURL:       api.URL,
		AuthURL:      auth.URL,
		ClientID:     "cid",
		ClientSecret: "csecret",
		OrgID:        "org",
		Environment:  "production",
		CacheTTL:     1 * time.Nanosecond, // force the GetValue read to bypass cache
	})
	require.NoError(t, err)

	// Sleep so the 1ns initial-fetch cache entry expires, forcing a fresh GetValue.
	time.Sleep(2 * time.Millisecond)

	v, ok, err := h.SecretConfig.Get("stripeApiKey")
	require.NoError(t, err)
	assert.True(t, ok)
	assert.Equal(t, "sk_live_refreshed", v)
	// Token minted twice: once initially, once after the 401 invalidation.
	assert.GreaterOrEqual(t, atomic.LoadInt32(&tokenMints), int32(2))
}

// --- §4: Health healthy / unhealthy ------------------------------------------

func TestHealth_HealthyAfterInit(t *testing.T) {
	h, _ := newHandle(t, map[string]any{"stripeApiKey": "x"})
	hh := h.Health()
	assert.Equal(t, "healthy", hh.Status)
	assert.True(t, hh.IsHealthy())

	// Free-function form.
	assert.Equal(t, "healthy", ConfigHealthOf(h).Status)
}

func TestHealth_UnhealthyWhenInitialFetchFails(t *testing.T) {
	stub := newStubClient(map[string]any{})
	stub.allValueErr = errors.New("config server unreachable")
	_, err := InitContainerConfig(context.Background(), InitContainerConfigOptions{
		Schema:       testSchema(t),
		ConfigClient: stub,
		EnvOverride:  map[string]string{"SMOOAI_CONFIG_ENV": "production"},
	})
	// Initial fetch failure surfaces at init (fail-loud).
	require.Error(t, err)
	assert.Contains(t, err.Error(), "config server unreachable")
}

func TestHealth_UnhealthyAfterRefreshFailurePastTTL(t *testing.T) {
	h, stub := newHandle(t, map[string]any{"stripeApiKey": "x"}, func(o *InitContainerConfigOptions) {
		o.CacheTTL = 5 * time.Millisecond
	})
	require.Equal(t, "healthy", h.Health().Status)

	// A refresh fails — within TTL we still report healthy (serve last-good).
	stub.getValueErr = errors.New("transient")
	_, _, _ = h.SecretConfig.Get("stripeApiKey")
	assert.Equal(t, "healthy", h.Health().Status, "within TTL a failed refresh stays healthy")

	// Past hard TTL, a failed refresh flips unhealthy.
	time.Sleep(8 * time.Millisecond)
	hh := h.Health()
	assert.Equal(t, "unhealthy", hh.Status)
	assert.Contains(t, hh.Reason, "transient")
}

func TestConfigHealthOf_NilHandle(t *testing.T) {
	hh := ConfigHealthOf(nil)
	assert.Equal(t, "unhealthy", hh.Status)
	assert.NotEmpty(t, hh.Reason)
}

func TestClient_EscapeHatch(t *testing.T) {
	h, stub := newHandle(t, map[string]any{})
	assert.Same(t, ConfigClient(stub), h.Client())
}

func TestEnvVarNameFor(t *testing.T) {
	cases := map[string]string{
		"stripeApiKey": "STRIPE_API_KEY",
		"apiUrl":       "API_URL",
		"betaEnabled":  "BETA_ENABLED",
		"x":            "X",
	}
	for in, want := range cases {
		assert.Equal(t, want, envVarNameFor(in), "envVarNameFor(%q)", in)
	}
}

File: src/container/__tests__/container.test.ts
import { defineConfig, StringSchema } from '@/config/config';
import { ConfigClient } from '@/platform/client';
import { TokenProvider } from '@/platform/TokenProvider';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { __resetSelectModeLogForTests, ConfigBootstrapError, configHealth, ConfigKeyUnresolvedError, initContainerConfig, selectMode } from '../index';

// Mock @smooai/fetch so ConfigClient/TokenProvider never hit the network.
const { mockFetch } = vi.hoisted(() => ({ mockFetch: vi.fn() }));
vi.mock('@smooai/fetch', () => ({ default: mockFetch }));

// A minimal schema with one secret + one public + one flag.
const schema = defineConfig({
    publicConfigSchema: { apiBaseUrl: StringSchema },
    secretConfigSchema: { stripeApiKey: StringSchema, sendgridApiKey: StringSchema },
    featureFlagSchema: { newCheckout: StringSchema },
});

/** A TokenProvider that returns a fixed JWT without an OAuth round-trip. */
class StubTokenProvider extends TokenProvider {
    public invalidateCallCount = 0;
    constructor(private readonly token: string = 'test-token') {
        super({ authUrl: 'https://stub.invalid', clientId: 'id', clientSecret: 'secret' });
    }
    async getAccessToken(): Promise<string> {
        return this.token;
    }
    invalidate(): void {
        this.invalidateCallCount++;
        super.invalidate();
    }
}

function makeClient(tokenProvider?: TokenProvider): ConfigClient {
    return new ConfigClient({
        baseUrl: 'https://api.smooai.test',
        orgId: 'org-1',
        environment: 'production',
        cacheTtlMs: 30_000,
        tokenProvider: tokenProvider ?? new StubTokenProvider(),
    });
}

/** Queue mock fetch responses (FIFO). */
function queueFetch(responses: Array<{ ok?: boolean; status?: number; json?: unknown; text?: string }>) {
    mockFetch.mockReset();
    for (const r of responses) {
        const ok = r.ok ?? true;
        const status = r.status ?? (ok ? 200 : 500);
        mockFetch.mockResolvedValueOnce({
            ok,
            status,
            statusText: ok ? 'OK' : 'ERR',
            json: async () => r.json,
            text: async () => r.text ?? JSON.stringify(r.json),
        });
    }
}

const ORIGINAL_ENV = { ...process.env };
// Schema-key env names that the env tier would read. Cleared per-test so a
// host shell that happens to export e.g. SENDGRID_API_KEY can't leak into the
// env tier and break isolation.
const SCHEMA_KEY_ENV = ['STRIPE_API_KEY', 'SENDGRID_API_KEY', 'API_BASE_URL', 'NEW_CHECKOUT'];
function clearSmooEnv() {
    for (const k of Object.keys(process.env)) {
        if (k.startsWith('SMOOAI_') || k.startsWith('SMOO_CONFIG')) delete process.env[k];
    }
    for (const k of SCHEMA_KEY_ENV) delete process.env[k];
}

beforeEach(() => {
    clearSmooEnv();
    mockFetch.mockReset();
    __resetSelectModeLogForTests();
});

afterEach(() => {
    for (const k of Object.keys(process.env)) {
        if (!(k in ORIGINAL_ENV)) delete process.env[k];
    }
    for (const [k, v] of Object.entries(ORIGINAL_ENV)) process.env[k] = v;
});

describe('initContainerConfig — bootstrap validation', () => {
    it('throws ConfigBootstrapError listing every missing required env var', async () => {
        // No env set, no injected client.
        const err = await initContainerConfig({ schema }).catch((e) => e);
        expect(err).toBeInstanceOf(ConfigBootstrapError);
        expect(err.missing).toEqual(
            expect.arrayContaining([
                'SMOOAI_CONFIG_API_URL',
                'SMOOAI_CONFIG_CLIENT_ID',
                'SMOOAI_CONFIG_CLIENT_SECRET',
                'SMOOAI_CONFIG_ORG_ID',
                'SMOOAI_CONFIG_ENV',
            ]),
        );
        expect(err.message).toContain('SMOOAI_CONFIG_API_URL');
    });

    it('lists only the actually-missing vars (partial env)', async () => {
        process.env.SMOOAI_CONFIG_API_URL = 'https://api.smooai.test';
        process.env.SMOOAI_CONFIG_CLIENT_ID = 'id';
        process.env.SMOOAI_CONFIG_ORG_ID = 'org-1';
        process.env.SMOOAI_CONFIG_ENV = 'production';
        // CLIENT_SECRET missing.
        const err = await initContainerConfig({ schema }).catch((e) => e);
        expect(err).toBeInstanceOf(ConfigBootstrapError);
        expect(err.missing).toEqual(['SMOOAI_CONFIG_CLIENT_SECRET']);
    });

    it('treats a blank (whitespace) env var as missing', async () => {
        process.env.SMOOAI_CONFIG_API_URL = 'https://api.smooai.test';
        process.env.SMOOAI_CONFIG_CLIENT_ID = '   ';
        process.env.SMOOAI_CONFIG_CLIENT_SECRET = 'secret';
        process.env.SMOOAI_CONFIG_ORG_ID = 'org-1';
        process.env.SMOOAI_CONFIG_ENV = 'production';
        const err = await initContainerConfig({ schema }).catch((e) => e);
        expect(err).toBeInstanceOf(ConfigBootstrapError);
        expect(err.missing).toEqual(['SMOOAI_CONFIG_CLIENT_ID']);
    });

    it('accepts legacy SMOOAI_CONFIG_API_KEY as the client secret', async () => {
        process.env.SMOOAI_CONFIG_API_URL = 'https://api.smooai.test';
        process.env.SMOOAI_CONFIG_CLIENT_ID = 'id';
        process.env.SMOOAI_CONFIG_API_KEY = 'legacy-secret';
        process.env.SMOOAI_CONFIG_ORG_ID = 'org-1';
        process.env.SMOOAI_CONFIG_ENV = 'production';
        // token mint + initial getAllValues
        queueFetch([{ json: { access_token: 'T', expires_in: 3600 } }, { json: { values: {} } }]);
        const handle = await initContainerConfig({ schema });
        expect(handle).toBeDefined();
    });

    it('with an injected ConfigClient, only SMOOAI_CONFIG_ENV is required', async () => {
        const err = await initContainerConfig({ schema, configClient: makeClient() }).catch((e) => e);
        expect(err).toBeInstanceOf(ConfigBootstrapError);
        expect(err.missing).toEqual(['SMOOAI_CONFIG_ENV']);
    });
});

describe('initContainerConfig — startup fetch (fail at boot, not first read)', () => {
    it('throws when the initial config fetch fails (no silent degraded start)', async () => {
        // initial getAllValues -> 500
        queueFetch([{ ok: false, status: 500, text: 'boom' }]);
        const err = await initContainerConfig({
            schema,
            environment: 'production',
            configClient: makeClient(),
        }).catch((e) => e);
        expect(err).toBeInstanceOf(Error);
        expect(String(err)).toMatch(/500/);
    });

    it('happy path: initial fetch succeeds and a value reads from cache without a 2nd HTTP call', async () => {
        // initial getAllValues returns the full map.
        queueFetch([{ json: { values: { stripeApiKey: 'sk_live_123', apiBaseUrl: 'https://x' } } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        const callsAfterInit = mockFetch.mock.calls.length;
        expect(await handle.secretConfig.get('stripeApiKey')).toBe('sk_live_123');
        expect(await handle.publicConfig.get('apiBaseUrl')).toBe('https://x');
        // getAllValues seeded the cache, so no extra fetches.
        expect(mockFetch.mock.calls.length).toBe(callsAfterInit);
    });
});

describe('fail-loud reads (§3)', () => {
    it('required secret unresolved -> throws ConfigKeyUnresolvedError (not undefined)', async () => {
        // initial getAllValues: empty. Then per-key getValue for the missing key -> {value: undefined}.
        queueFetch([{ json: { values: {} } }, { json: { value: undefined } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        const err = await handle.secretConfig.get('stripeApiKey').catch((e) => e);
        expect(err).toBeInstanceOf(ConfigKeyUnresolvedError);
        expect(err.key).toBe('stripeApiKey');
        expect(err.env).toBe('production');
        expect(err.triedTiers).toEqual(['env', 'http']);
    });

    it('optional key absent -> returns undefined, does NOT throw', async () => {
        queueFetch([{ json: { values: {} } }, { json: { value: undefined } }]);
        const handle = await initContainerConfig({
            schema,
            environment: 'production',
            configClient: makeClient(),
            optionalKeys: ['sendgridApiKey'],
        });
        await expect(handle.secretConfig.get('sendgridApiKey')).resolves.toBeUndefined();
    });

    it('getSync for an unresolved required key throws (no silent undefined)', async () => {
        queueFetch([{ json: { values: {} } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        expect(() => handle.secretConfig.getSync('stripeApiKey')).toThrow(ConfigKeyUnresolvedError);
    });

    it('getSync returns a cached value when present', async () => {
        queueFetch([{ json: { values: { stripeApiKey: 'sk_cached' } } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        expect(handle.secretConfig.getSync('stripeApiKey')).toBe('sk_cached');
    });

    it('an explicit process env override wins over HTTP (env tier precedence)', async () => {
        process.env.STRIPE_API_KEY = 'sk_from_env';
        queueFetch([{ json: { values: { stripeApiKey: 'sk_from_http' } } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        expect(await handle.secretConfig.get('stripeApiKey')).toBe('sk_from_env');
    });
});

describe('401 -> refresh -> retry (§5)', () => {
    it('invalidates the token and retries once on a 401', async () => {
        const tp = new StubTokenProvider();
        const client = makeClient(tp);
        // initial getAllValues ok (empty)
        // then per-key getValue: 401 (triggers invalidate+retry), then 200.
        queueFetch([{ json: { values: {} } }, { ok: false, status: 401, text: 'expired' }, { json: { value: 'sk_after_refresh' } }]);
        // @smooai/fetch throws on non-2xx with err.response.status — emulate that
        // for the 401 by making the 2nd call reject.
        mockFetch.mockReset();
        mockFetch
            .mockResolvedValueOnce({ ok: true, status: 200, json: async () => ({ values: {} }) })
            .mockRejectedValueOnce(Object.assign(new Error('HTTP 401'), { response: { status: 401 } }))
            .mockResolvedValueOnce({ ok: true, status: 200, json: async () => ({ value: 'sk_after_refresh' }) });

        const handle = await initContainerConfig({ schema, environment: 'production', configClient: client });
        const v = await handle.secretConfig.get('stripeApiKey');
        expect(v).toBe('sk_after_refresh');
        expect(tp.invalidateCallCount).toBe(1);
    });
});

describe('configHealth (§4)', () => {
    it('reports healthy after a successful initial fetch', async () => {
        queueFetch([{ json: { values: { stripeApiKey: 'sk' } } }]);
        const handle = await initContainerConfig({ schema, environment: 'production', configClient: makeClient() });
        expect(handle.health()).toEqual({ status: 'healthy' });
        expect(configHealth(handle)).toEqual({ status: 'healthy' });
    });

    it('serves healthy on a background refresh failure within TTL, unhealthy past hard-expiry', async () => {
        vi.useFakeTimers();
        try {
            // initial getAllValues seeds stripeApiKey.
            mockFetch.mockReset();
            mockFetch
                .mockResolvedValueOnce({ ok: true, status: 200, json: async () => ({ values: { stripeApiKey: 'sk_initial' } }) })
                // subsequent getValue refresh fails.
                .mockRejectedValue(Object.assign(new Error('network down'), { response: { status: 503 } }));

            const handle = await initContainerConfig({
                schema,
                environment: 'production',
                configClient: makeClient(),
                cacheTtlMs: 30_000,
            });
            expect(handle.health()).toEqual({ status: 'healthy' });

            // Force a refresh failure by reading after the per-key cache expires.
            // The cached value from getAllValues is still served (last-good).
            vi.advanceTimersByTime(31_000); // expire the per-key TTL cache
            const v = await handle.secretConfig.get('stripeApiKey').catch((e) => e);
            // last-good cache is gone (TTL expired) AND http failed -> unresolved required.
            expect(v).toBeInstanceOf(ConfigKeyUnresolvedError);

            // Health: last refresh failed and we're past the TTL window -> unhealthy.
            const h = handle.health();
            expect(h.status).toBe('unhealthy');
            if (h.status === 'unhealthy') expect(h.reason).toMatch(/network down|TTL/);
        } finally {
            vi.useRealTimers();
        }
    });
});

describe('selectMode (§2)', () => {
    it('SMOOAI_CONFIG_MODE=container forces container mode', () => {
        expect(selectMode({ mode: 'container' })).toBe('container');
        expect(selectMode({ mode: 'CONTAINER' })).toBe('container');
    });

    it('a present blob source means default mode (not container)', () => {
        expect(selectMode({ blobPresent: true, clientId: 'id', clientSecret: 's', apiUrl: 'u' })).toBe('default');
    });

    it('a present file source means default mode', () => {
        expect(selectMode({ filePresent: true, clientId: 'id', clientSecret: 's', apiUrl: 'u' })).toBe('default');
    });

    it('auto-selects container when CLIENT_ID + CLIENT_SECRET + API_URL set and no blob/file', () => {
        expect(selectMode({ clientId: 'id', clientSecret: 's', apiUrl: 'u' })).toBe('container');
    });

    it('falls back to default when the M2M creds are incomplete', () => {
        expect(selectMode({ clientId: 'id', apiUrl: 'u' })).toBe('default');
        expect(selectMode({})).toBe('default');
    });

    it('reads from process.env when inputs are omitted', () => {
        process.env.SMOOAI_CONFIG_MODE = 'container';
        expect(selectMode()).toBe('container');
    });
});
